rfcs/0011-separate-flyway-from-pipeline.md

```diff
@@ -0,0 +1,382 @@
+
+* Test infrastructure will be altered to mimic the deployment by running the new application prior to starting the local
+BFD server for tests that run the BFD server as a standalone process.
+
+* Test infrastructure will be altered to invoke the new application logic via method call prior to running tests that
+require a BFD database but do not launch a BFD server.
+
+* The new Java application code will reside in the existing BFD repo as another sub-module of bfd-parent so that it will
+be built and tested as part of the current Github actions application workflow and Jenkins App build stage.
+
+Deployment design:
+
+![Proposed Deployment Sequence](./resources/0011-proposed-deployment-sequence.svg)
+
+<details><summary>Diagram Notes</summary>
+Like the previous diagram, this illustrates the deployment sequence with some broad definitions. This adds the `bfd-db-migrator` participant that includes a blocking deployment stage for each environment that must complete **before** proceeding to apply updates to the `bfd-pipeline` and `bfd-server` resources for each environment.
+</details>
+
+* A new migrator service account and associated migrator database role that has all required privileges for running
+migrations and hibernate validation will be created.
+
+* The Jenkins APP AMI build phase will build a new migrator AMI that includes the new application.
+
+* The migrator AMI and migrator service account will be used to run the migrator application on a new EC2 instance as
+part of a new Jenkins pipeline stage that will be inserted prior to each of the three stages that deploy to TEST,
+PROD-SBX, PROD.
+
+* The Jenkins pipeline will not proceed to the next stage (deploying the other applications) until the migrator
+application has exited with a status indicating success. In the event that this application status indicates that not
+all operations were successful, the Jenkins deployment will halt and be marked as a failure.
+
+* The new application will produce log files that contain the Flyway and Hibernate logging that is currently captured in
+the BFD Pipeline application logs. These new application logs will be sent to Splunk and Cloudwatch.
+
+* The migrator EC2 instance could be powered off or terminated once the migrator application exits and all artifacts
+have been gathered.
+
+New Flyway migration once the above is in place:
+
+* A new Flyway migration will be developed that creates all non-user roles and sets privileges correctly on all existing
+database objects.
+
+### Proposed Solution: Unresolved Questions
+[Proposed Solution: Unresolved Questions]: #proposed-solution-unresolved-questions
+
+Do we need to continue to invoke Hibernate validation at all? What benefit do we derive from this check that is not
+already derived from running the unit and integration tests? What sorts of problems would not be caught by the tests
+that would be caught by hibernate validation?
+
```

Comment:
Invoking Hibernate validation in the migrator app takes very little time and provides a sanity check to ensure the migration did not fail in some odd way that the app did not otherwise detect.